File: python-scraper/services/normalize.py
"""Cleaning + normalization helpers. Mirrors normalizeEvent() from the
Node backend so both services produce compatible rows."""
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any, Iterable, Optional
from urllib.parse import urlparse

from models.raw_event import RawEvent

KNOWN_TYPES = {"hackathon", "webinar", "workshop"}
KNOWN_MODES = {"online", "offline", "hybrid"}


def parse_date(value: Any) -> Optional[datetime]:
    if value is None:
        return None
    if isinstance(value, datetime):
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value
    if isinstance(value, (int, float)):
        # Heuristic: treat as ms if it's clearly milliseconds
        try:
            ts = float(value)
            if ts > 1e12:
                ts = ts / 1000.0
            return datetime.fromtimestamp(ts, tz=timezone.utc)
        except (OverflowError, ValueError, OSError):
            return None
    if isinstance(value, str):
        s = value.strip()
        if not s:
            return None
        # Try ISO 8601 first
        try:
            # Handle trailing Z
            iso = s.replace("Z", "+00:00")
            dt = datetime.fromisoformat(iso)
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            return dt
        except ValueError:
            pass
        # Try a few common formats
        for fmt in (
            "%Y-%m-%d %H:%M:%S",
            "%Y-%m-%d",
            "%d %b %Y",
            "%d %B %Y",
            "%b %d, %Y",
            "%B %d, %Y",
        ):
            try:
                dt = datetime.strptime(s, fmt)
                return dt.replace(tzinfo=timezone.utc)
            except ValueError:
                continue
    return None


def normalize_type(value: Optional[str]) -> str:
    v = (value or "").lower().strip()
    if v in KNOWN_TYPES:
        return v
    if "hack" in v:
        return "hackathon"
    if "webinar" in v:
        return "webinar"
    if "workshop" in v:
        return "workshop"
    return "other"


def normalize_mode(value: Optional[str]) -> str:
    v = (value or "").lower().strip()
    if v in KNOWN_MODES:
        return v
    if any(k in v for k in ("online", "virtual", "remote")):
        return "online"
    if any(k in v for k in ("offline", "in-person", "in person", "onsite", "on-site")):
        return "offline"
    if "hybrid" in v:
        return "hybrid"
    return "unknown"


def is_valid_url(url: Optional[str]) -> bool:
    if not url:
        return False
    try:
        parsed = urlparse(url)
        return parsed.scheme in ("http", "https") and bool(parsed.netloc)
    except Exception:
        return False


def clean_tags(tags: Optional[Iterable[str]]) -> list[str]:
    if not tags:
        return []
    seen: dict[str, None] = {}
    for t in tags:
        if not isinstance(t, str):
            continue
        s = re.sub(r"\s+", " ", t).strip()
        if 0 < len(s) < 60 and s.lower() not in seen:
            seen[s.lower()] = None
            if len(seen) >= 12:
                break
    return [k for k in seen.keys()]


def normalize_event(raw: RawEvent, source: str) -> Optional[dict]:
    title = (raw.title or "").strip()
    url = (raw.url or "").strip()
    if not title or not is_valid_url(url):
        return None

    end = parse_date(raw.end_date)
    start = parse_date(raw.start_date)

    return {
        "title": title[:500],
        "platform": source,
        "type": normalize_type(raw.type),
        "url": url,
        "image": (raw.image or "").strip() or None,
        "start_date": start,
        "end_date": end,
        "mode": normalize_mode(raw.mode),
        "tags": clean_tags(raw.tags),
        "organizer": (raw.organizer or "").strip() or None,
        "location": (raw.location or "").strip() or None,
        "prize": (raw.prize or "").strip() or None,
        "description": (raw.description or "").strip() or None,
    }

File: artifacts/api-server/src/lib/normalize.ts
import type { InsertEvent } from "@workspace/db";

export type RawEvent = {
  title?: string | null;
  url?: string | null;
  image?: string | null;
  startDate?: string | Date | null;
  endDate?: string | Date | null;
  deadline?: string | Date | null;
  mode?: string | null;
  type?: string | null;
  tags?: string[] | null;
  organizer?: string | null;
  location?: string | null;
  prize?: string | null;
  description?: string | null;
};

const KNOWN_TYPES = new Set(["hackathon", "webinar", "workshop"]);
const KNOWN_MODES = new Set(["online", "offline", "hybrid"]);

function parseDate(value: string | Date | null | undefined): Date | null {
  if (!value) return null;
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? null : value;
  }
  const trimmed = value.trim();
  if (!trimmed) return null;
  const d = new Date(trimmed);
  if (!Number.isNaN(d.getTime())) return d;
  return null;
}

function normalizeType(value?: string | null): InsertEvent["type"] {
  const v = (value ?? "").toLowerCase().trim();
  if (KNOWN_TYPES.has(v)) return v;
  if (v.includes("hack")) return "hackathon";
  if (v.includes("webinar")) return "webinar";
  if (v.includes("workshop")) return "workshop";
  return "other";
}

function normalizeMode(value?: string | null): InsertEvent["mode"] {
  const v = (value ?? "").toLowerCase().trim();
  if (KNOWN_MODES.has(v)) return v;
  if (v.includes("online") || v.includes("virtual") || v.includes("remote"))
    return "online";
  if (v.includes("offline") || v.includes("in-person") || v.includes("onsite"))
    return "offline";
  if (v.includes("hybrid")) return "hybrid";
  return "unknown";
}

export function normalizeEvent(
  raw: RawEvent,
  source: string,
): InsertEvent | null {
  const title = raw.title?.trim();
  const url = raw.url?.trim();
  if (!title || !url) return null;

  const end = parseDate(raw.endDate ?? raw.deadline);
  const start = parseDate(raw.startDate);

  const tags = (raw.tags ?? [])
    .map((t) => t.trim())
    .filter((t) => t.length > 0 && t.length < 60)
    .slice(0, 12);

  return {
    title: title.slice(0, 500),
    platform: source,
    type: normalizeType(raw.type),
    url,
    image: raw.image?.trim() || null,
    startDate: start,
    endDate: end,
    mode: normalizeMode(raw.mode),
    tags,
    organizer: raw.organizer?.trim() || null,
    location: raw.location?.trim() || null,
    prize: raw.prize?.trim() || null,
    description: raw.description?.trim() || null,
  };
}
